### PR TITLE
(TEPHRA-215) (TEPHRA-218) Use single thread across all regions in a r…

### PR DESCRIPTION
**Pull request by gokulavasan**
…egion server to persist Prune Upper Bound info. Also don't refresh cache during startup of TransactionStateCache to avoid the possibility of Service stopping if tx.snapshot dir is not found during startup

JIRA : 
https://issues.apache.org/jira/browse/TEPHRA-215
https://issues.apache.org/jira/browse/TEPHRA-218

### CODE REVIEW

File: tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplier.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * Supplies instances of {@link PruneUpperBoundWriter} implementations.
+ */
+public class PruneUpperBoundWriterSupplier implements Supplier<PruneUpperBoundWriter> {
+
+  private final DataJanitorState dataJanitorState;
+  private final long pruneFlushInterval;
+
+  protected static volatile PruneUpperBoundWriter instance;
```
Comment (poornachandra) [inline tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplier.java]: I don't think this needs to be protected

File: tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplier.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  protected static Object lock = new Object();
```
Comment (poornachandra) [inline tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplier.java]: Same here about protected. Also it is better to make the `lock` object final.

File: tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriter.java [hunk 3]
```diff
@@ -18,55 +18,48 @@
...
-    this.pruneStateTable = pruneStateTable;
+  public PruneUpperBoundWriter(DataJanitorState dataJanitorState, long pruneFlushInterval) {
     this.dataJanitorState = dataJanitorState;
-    this.regionName = regionName;
-    this.regionNameAsString = regionNameAsString;
     this.pruneFlushInterval = pruneFlushInterval;
-    this.pruneUpperBound = new AtomicLong();
-    this.shouldFlush = new AtomicBoolean(false);
+    this.pruneEntries = new ConcurrentLinkedQueue<>();
     startFlushThread();
```
Comment (poornachandra) [inline tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriter.java]: Since the flush thread is started in the `startUp` method, we don't need to start it in the constructor.

File: tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriter.java [hunk 4]
```diff
@@ -18,55 +18,48 @@
...
   }
 
-  public boolean isAlive() {
-    return flushThread.isAlive();
+  public void persistPruneEntry(byte[] regionName, long pruneUpperBound) {
+    pruneEntries.add(new PruneInfo(regionName, pruneUpperBound));
```
Comment (poornachandra) [inline tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriter.java]: I think it would be good to limit the number of entries in the `pruneEntries` queue. If we are over the limit then we can drop earlier elements.

File: tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriter.java [hunk 5]
```diff
@@ -79,15 +72,17 @@ public void run() {
         while (!isInterrupted()) {
           long now = System.currentTimeMillis();
           if (now > (lastChecked + pruneFlushInterval)) {
-            if (shouldFlush.compareAndSet(true, false)) {
+            if (!pruneEntries.isEmpty()) {
               // should flush data
               try {
-                dataJanitorState.savePruneUpperBoundForRegion(regionName, pruneUpperBound.get());
+                Iterator<PruneInfo> iterator = pruneEntries.iterator();
```
Comment (poornachandra) [inline tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriter.java]: Since we are doing concurrent operations here it would be good to use `peek()` and `poll()`, rather than a iterator.

File: tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplier.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    this.pruneFlushInterval = pruneFlushInterval;
+  }
+
+  @Override
+  public PruneUpperBoundWriter get() {
+    if (instance == null) {
+      synchronized (lock) {
+        if (instance == null) {
+          instance = new PruneUpperBoundWriter(dataJanitorState, pruneFlushInterval);
+          instance.start();
```
Comment (poornachandra) [inline tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplier.java]: We have the same problem as `TransactionStateCache` here too - we don't have a way to stop this thread after all instances of `TransactionProcessor` co-processors have been shutdown (https://issues.apache.org/jira/browse/TEPHRA-152)

Comment (gokulavasan) [inline tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplier.java]: @poornachandra Do you have any thoughts on how to address this?

File: tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriter.java (hunk 4 above)
Comment (gokulavasan) [inline tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriter.java]: @poornachandra Should that max size be configurable?

**Comment by gokulavasan** [comment]
@poornachandra Can you please take a look again. Thanks!

**Comment by gokulavasan** [comment]
@poornachandra Please review when you get a chance. Thanks!

File: tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriter.java [hunk 7]
```diff
@@ -18,56 +18,56 @@
...
+  private static final Log LOG = LogFactory.getLog(PruneUpperBoundWriter.class);
 
-  private final TableName pruneStateTable;
   private final DataJanitorState dataJanitorState;
-  private final byte[] regionName;
-  private final String regionNameAsString;
   private final long pruneFlushInterval;
-  private final AtomicLong pruneUpperBound;
-  private final AtomicBoolean shouldFlush;
+  private final ConcurrentSkipListMap<ByteBuffer, Long> pruneEntries;
```
Comment (poornachandra) [inline tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriter.java]: It would be simpler to use `ConcurrentSkipListMap<byte[], Long>`

File: tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriter.java [hunk 8]
```diff
@@ -76,18 +76,21 @@ private void startFlushThread() {
...
+                  // We can now remove the entry only if the key and value match with what we wrote since it is
+                  // possible that a new pruneUpperBound for the same key has been added
+                  pruneEntries.remove(firstEntry.getKey(), firstEntry.getValue());
+                }
               } catch (IOException ex) {
-                LOG.warn("Cannot record prune upper bound for region " + regionNameAsString + " in the table " +
-                           pruneStateTable.getNameWithNamespaceInclAsString() + " after compacting region.", ex);
-                // Retry again
-                shouldFlush.set(true);
+                LOG.warn("Cannot record prune upper bound for a region.", ex);
```
Comment (poornachandra) [inline tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriter.java]: It would be good to add prune state table name to the log message, so that it can help in figuring out mis-configurations if any

File: tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplier.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+  private final long pruneFlushInterval;
+
+  public PruneUpperBoundWriterSupplier(DataJanitorState dataJanitorState, long pruneFlushInterval) {
+    this.dataJanitorState = dataJanitorState;
+    this.pruneFlushInterval = pruneFlushInterval;
+  }
+
+  @Override
+  public PruneUpperBoundWriter get() {
+    synchronized (lock) {
```
Comment (poornachandra) [inline tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplier.java]: We can make an `instance == null` check outside the lock. In case instance is not null just increment refCount and return.

File: tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplier.java [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+        instance.startAndWait();
+      }
+      refCount++;
+      return instance;
+    }
+  }
+
+  public void release() {
+    synchronized (lock) {
+      refCount--;
```
Comment (poornachandra) [inline tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplier.java]: Same here, decrement refCount outside lock and return if refCount is greater than zero. Only lock when stopping.

File: tephra-hbase-compat-1.1-base/src/test/java/org/apache/tephra/hbase/txprune/InvalidListPruneTest.java [hunk 11]
```diff
@@ -360,8 +404,18 @@ public TransactionStateCache get() {
...
+      synchronized (compactionLock) {
+        super.postCompact(e, store, resultFile, request);
+        lastMajorCompactionTime.set(System.currentTimeMillis());
+      }
+    }
+
+    @Override
+    public void stop(CoprocessorEnvironment e) throws IOException {
+      synchronized (stopLock) {
+        System.out.println("*********** Stopping " + this.getClass().getName());
```
Comment (poornachandra) [inline tephra-hbase-compat-1.1-base/src/test/java/org/apache/tephra/hbase/txprune/InvalidListPruneTest.java]: Remove the debug println

File: tephra-hbase-compat-1.1-base/src/test/java/org/apache/tephra/hbase/txprune/InvalidListPruneTest.java [hunk 12]
```diff
@@ -334,6 +337,45 @@ public Table get() throws IOException {
...
 
+  @Test(timeout = 60000L)
+  public void testClusterShutdown() throws Exception {
+    java.util.concurrent.ExecutorService executorService = Executors.newSingleThreadExecutor();
+    try {
+      // Create a new transaction snapshot
+      InMemoryTransactionStateCache.setTransactionSnapshot(
+        new TransactionSnapshot(100, 100, 100, ImmutableSet.of(50L),
+                                ImmutableSortedMap.<Long, TransactionManager.InProgressTx>of()));
+      // Run major compaction
```
Comment (poornachandra) [inline tephra-hbase-compat-1.1-base/src/test/java/org/apache/tephra/hbase/txprune/InvalidListPruneTest.java]: The comment can be changed to - `Try to create a situation where HBase shuts down when the async prune writer thread is trying to write to the prune table on a compaction. This will surface any deadlocks during HBase shutdown.`

File: tephra-hbase-compat-1.1-base/src/test/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplierTest.java [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Tests to verify the behavior of {@link PruneUpperBoundWriterSupplier}.
+ */
+public class PruneUpperBoundWriterSupplierTest extends AbstractHBaseTableTest {
```
Comment (poornachandra) [inline tephra-hbase-compat-1.1-base/src/test/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplierTest.java]: Can you add comments on what this test is trying to do? I'm not sure I follow the test.

Also, do we need to start an HBase instance for this test? We just need to make sure that prune writer gets stopped when reference count reaches zero, right? In that case can we pass in DataJanitorState as null to PruneUpperBoundWriterSupplier, and not call persistPruneEntry?

File: tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriter.java [hunk 14]
```diff
@@ -18,56 +18,56 @@
...
-    this.shouldFlush.set(true);
+  @Override
+  protected void startUp() throws Exception {
+    startFlushThread();
   }
 
-  public void stop() {
+  @Override
+  protected void shutDown() throws Exception {
     if (flushThread != null) {
```
Comment (poornachandra) [inline tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriter.java]: Also would be good to add a log statement saying "stopping PruneUpperBoundWriter" and "stopped PruneUpperBoundWriter"

File: tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriter.java (hunk 7 above)
Comment (gokulavasan) [inline tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriter.java]: We need the key to be Comparable. Hence we need to use ByteBuffer.

File: tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriter.java [hunk 15]
```diff
@@ -18,56 +18,56 @@
...
+      stopped = true;
       flushThread.interrupt();
     }
```
Comment (gokulavasan) [inline tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriter.java]: What would be a good timeout?

**Comment by gokulavasan** [comment]
@poornachandra Addressed comments. Please take a look when you get a chance.

File: tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriter.java (hunk 14 above)
Comment (gokulavasan) [inline tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriter.java]: Added log statements.

File: tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplier.java [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+
+  @Override
+  public PruneUpperBoundWriter get() {
+    synchronized (lock) {
+      if (instance == null) {
+        instance = new PruneUpperBoundWriter(tableName, dataJanitorState, pruneFlushInterval);
+        instance.startAndWait();
+      }
+      refCount++;
+      LOG.debug("Incrementing Reference Count for PruneUpperBoundWriter " + refCount);
```
Comment (poornachandra) [inline tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplier.java]: Would be good to enclose this log statement with `LOG.isDebugEnabled()`

File: tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplier.java [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+      return instance;
+    }
+  }
+
+  public void release() {
+    synchronized (lock) {
+      refCount--;
+      LOG.debug("Decrementing Reference Count for PruneUpperBoundWriter " + refCount);
```
Comment (poornachandra) [inline tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplier.java]: Same here

File: tephra-hbase-compat-1.1-base/src/test/java/org/apache/tephra/hbase/txprune/InvalidListPruneTest.java [hunk 18]
```diff
@@ -334,6 +337,46 @@ public Table get() throws IOException {
     }
   }
 
+  @Test(timeout = 60000L)
+  public void testClusterShutdown() throws Exception {
```
Comment (poornachandra) [inline tephra-hbase-compat-1.1-base/src/test/java/org/apache/tephra/hbase/txprune/InvalidListPruneTest.java]: Is this test required now?

File: tephra-hbase-compat-1.1-base/src/test/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplierTest.java [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Tests to verify the behavior of {@link PruneUpperBoundWriterSupplier}.
+ */
+public class PruneUpperBoundWriterSupplierTest {
+  private static final int NUM_OPS = 10;
```
Comment (poornachandra) [inline tephra-hbase-compat-1.1-base/src/test/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplierTest.java]: Since this is a fast operation, let's make this a bigger number. 1000 maybe?

File: tephra-hbase-compat-1.1-base/src/test/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplierTest.java [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+    final PruneUpperBoundWriterSupplier supplier = new PruneUpperBoundWriterSupplier(null, null, 10L);
+    final PruneUpperBoundWriter writer = supplier.get();
+    final AtomicInteger numOps = new AtomicInteger(NUM_OPS);
+    final Random random = new Random(System.currentTimeMillis());
+
+    // Start threads that will 'get' PruneUpperBoundWriters
+    ExecutorService executor = Executors.newFixedThreadPool(3);
+    List<Future> futureList = new ArrayList<>();
+    for (int i = 0; i < 3; i++) {
+      futureList.add(executor.submit(new Callable<Void>() {
```
Comment (poornachandra) [inline tephra-hbase-compat-1.1-base/src/test/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplierTest.java]: `Callable` can be replaced with `Runnable` so that the return statement is not required.

File: tephra-hbase-compat-1.1-base/src/test/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplierTest.java [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+    }
+    executor.shutdown();
+    executor.awaitTermination(2, TimeUnit.SECONDS);
+
+    futureList.clear();
+    numOps.set(NUM_OPS);
+    // Start thread that release PruneUpperBoundWriters
+    executor = Executors.newFixedThreadPool(3);
+    for (int i = 0; i < 3; i++) {
+      futureList.add(executor.submit(new Callable<Void>() {
```
Comment (poornachandra) [inline tephra-hbase-compat-1.1-base/src/test/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplierTest.java]: Same here about using `Runnable`

File: tephra-hbase-compat-1.1-base/src/test/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplierTest.java [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+  public void testSupplier() throws Exception {
+    final PruneUpperBoundWriterSupplier supplier = new PruneUpperBoundWriterSupplier(null, null, 10L);
+    final PruneUpperBoundWriter writer = supplier.get();
+    final AtomicInteger numOps = new AtomicInteger(NUM_OPS);
+    final Random random = new Random(System.currentTimeMillis());
+
+    // Start threads that will 'get' PruneUpperBoundWriters
+    ExecutorService executor = Executors.newFixedThreadPool(3);
+    List<Future> futureList = new ArrayList<>();
+    for (int i = 0; i < 3; i++) {
```
Comment (poornachandra) [inline tephra-hbase-compat-1.1-base/src/test/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplierTest.java]: It would be good to have more threads too

File: tephra-hbase-compat-1.1-base/src/test/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplierTest.java [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+    // Start thread that release PruneUpperBoundWriters
+    executor = Executors.newFixedThreadPool(3);
+    for (int i = 0; i < 3; i++) {
+      futureList.add(executor.submit(new Callable<Void>() {
+
+        @Override
+        public Void call() throws Exception {
+          // We need to release all NUM_OPS 'gets' that were executed to trigger shutdown of the single instance of
+          // PruneUpperBoundWriter
+          while (numOps.decrementAndGet() >= 0) {
```
Comment (poornachandra) [inline tephra-hbase-compat-1.1-base/src/test/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplierTest.java]: Can we change this to leave one instance out, so that we can assert that that instance is still running after the n-1 instances were released back. We can later release it separately and then assert that it is not running.

**Comment by gokulavasan** [comment]
@poornachandra Addressed comments. Please take a look when you get a chance.

File: tephra-hbase-compat-1.1-base/src/test/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplierTest.java [hunk 24]
```diff
@@ -74,30 +81,35 @@ public Void call() throws Exception {
...
-          return null;
         }
       }));
     }
 
     for (Future future : futureList) {
       future.get(1, TimeUnit.SECONDS);
     }
+    // Since we got one instance in the beginning, we need to release it
+    supplier.release();
```
Comment (poornachandra) [inline tephra-hbase-compat-1.1-base/src/test/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriterSupplierTest.java]: It would be good to assert that the writer is running and alive before we do the final release

**Comment by gokulavasan** [comment]
@poornachandra Ported changes to other compat modules. Please review when you get a chance.

**Comment by poornachandra** [comment]
LGTM
